File: python_port/feature_extraction/pick_reference_phase.py
"""Port of matlab/feature_extraction/pick_reference_phase.m.

Chooses a reference phase from `TestBrake` (the output of
`braking_detection.detect_braking_struct_beta`) only if ALL expected BC and
WV streams are valid (GPS ignored). Pure query function -- reads TestBrake,
adds no fields to it.

Eligibility (per phase):
  MBP: Time & Pressure nonempty, equal length, >= 10 samples
  BC : #valid == nBC_expected, each has Time/Pressure nonempty,
       SensorError == 0, NormalBraking == 1
  WV : #valid == nWV_expected, each has Time/Pressure nonempty,
       WV_SensorError == 0

Bug fix, documented (not a silent replication): the MATLAB source checks an
*optional* field named `'BrakingAct'` for BC validity, which does not exist
anywhere in this codebase -- confirmed by reading every field the sibling
files (`detect_braking_struct_beta.m`, `build_TestBrake_sets.m`,
`Collect_Healthy_SensorData.m`) actually set/read; they all use
`NormalBraking` for this exact concept, and `Collect_Healthy_SensorData.m`'s
own docstring even still says "BrakingAct==1" while its real code already
reads `.NormalBraking` -- strong evidence `BrakingAct` is a stale, unrenamed
field-name string, not intentional. Because `count_valid_streams()`'s
optional-field check defaults to "pass" when the field is absent, the
*effect* of this bug in the original is not "BC validity always fails" --
it's "the NormalBraking condition is silently never enforced," a weaker
filter than documented. This port uses `'NormalBraking'`, restoring the
documented behavior.
"""
from __future__ import annotations

import math
from typing import Iterable, Optional

import numpy as np


def _has_nonempty_series(phase: dict, t_field: str, p_field: str, min_len: int) -> bool:
    t = phase.get(t_field)
    p = phase.get(p_field)
    if t is None or p is None or len(t) == 0 or len(p) == 0:
        return False
    return len(t) == len(p) and len(t) >= min_len


def _all_equal(value, target) -> bool:
    """Port of MATLAB's all(X.(field)==val). Fields checked here are always
    plain scalar bools in this codebase's BC/WV dicts, but np.all() is used
    for robustness, matching MATLAB's array-safe `all`."""
    return bool(np.all(np.asarray(value) == target))


def _count_valid_streams(
    entries: Optional[list], req_field: str, req_val, opt_field: Optional[str] = None, opt_val=0
) -> tuple:
    """Port of count_valid_streams(A,tName,pName,reqField,reqVal,optField,optVal).
    tName/pName are always 'Time'/'Pressure' at both call sites in the
    source, so hardcoded here rather than threaded through as parameters."""
    if not entries:
        return 0, 0
    n_total = len(entries)
    n_valid = 0
    for x in entries:
        has_t = x.get("Time") is not None and len(x["Time"]) > 0
        has_p = x.get("Pressure") is not None and len(x["Pressure"]) > 0
        req_ok = req_field in x and x[req_field] is not None and _all_equal(x[req_field], req_val)
        opt_ok = True
        if opt_field:
            if opt_field in x and x[opt_field] is not None:
                opt_ok = _all_equal(x[opt_field], opt_val)
        if has_t and has_p and req_ok and opt_ok:
            n_valid += 1
    return n_valid, n_total


def _any_true_field(phase: dict, fields: Iterable[str]) -> bool:
    for f in fields:
        v = phase.get(f)
        if v is None:
            continue
        if isinstance(v, (list, tuple, np.ndarray)) and len(v) == 0:
            continue
        if bool(np.any(np.asarray(v) != 0)):
            return True
    return False


def pick_reference_phase(test_brake: list) -> tuple:
    """Returns (ref_phase, scores, report).

    ref_phase: 1-indexed phase number (matching each phase's own 'PhaseIdx'
    field, and MATLAB's 1-based `TestBrake(refPhase)` indexing) of the
    highest-scoring eligible phase, or None if no phase is eligible
    (MATLAB's NaN).
    scores: list[float], one per phase, -inf for ineligible, 100-10*has_err
    for eligible (MATLAB ties broken by first occurrence; np.argmax matches).
    report: list[dict], one per phase, diagnostic fields.
    """
    n = len(test_brake)
    scores = [-math.inf] * n
    report: list = []

    if n == 0:
        return None, scores, report

    nbc_expected = max(len(s.get("BC") or []) for s in test_brake)
    nwv_expected = max(len(s.get("WV") or []) for s in test_brake)

    for s in test_brake:
        mbp_ok = _has_nonempty_series(s, "MBP_Time", "MBP_Pressure", 10)

        bc_valid, bc_total = _count_valid_streams(s.get("BC"), "SensorError", 0, "NormalBraking", 1)
        wv_valid, wv_total = _count_valid_streams(s.get("WV"), "WV_SensorError", 0)

        bc_ok = bc_valid == nbc_expected and bc_total >= nbc_expected
        wv_ok = wv_valid == nwv_expected and wv_total >= nwv_expected

        has_err = _any_true_field(s, ("SV_Error", "UP_Error", "EmergencyBrake"))

        i = len(report)
        if mbp_ok and bc_ok and wv_ok:
            scores[i] = 100 - 10 * has_err

        report.append({
            "MBP_OK": mbp_ok,
            "BC_valid": bc_valid, "BC_total": bc_total, "BC_expected": nbc_expected, "BC_OK": bc_ok,
            "WV_valid": wv_valid, "WV_total": wv_total, "WV_expected": nwv_expected, "WV_OK": wv_ok,
            "HasErrors": has_err,
        })

    scores_arr = np.asarray(scores, dtype=np.float64)
    if np.any(np.isfinite(scores_arr)):
        ref_phase = int(np.argmax(scores_arr)) + 1  # 1-indexed
    else:
        ref_phase = None

    return ref_phase, scores, report

File: matlab/feature_extraction/pick_reference_phase.m
function [refPhase, scores, report] = pick_reference_phase(TestBrake)
%PICK_REFERENCE_PHASE_STRICT  Choose a reference phase only if ALL expected
%BC and WV streams are valid (GPS ignored).
%
% Eligibility (per phase)
%   MBP: Time & Pressure nonempty, equal length, >= 10 samples
%   BC : #valid == nBC_expected, each has Time/Pressure nonempty,
%        SensorError == 0, BrakingAct == 1 (if field exists)
%   WV : #valid == nWV_expected, each has Time/Pressure nonempty,
%        WV_SensorError == 0
%
% Outputs
%   refPhase : index of chosen phase (NaN if none)
%   scores   : 100 for eligible, -Inf otherwise
%   report   : diagnostics

    N = numel(TestBrake);
    scores = -inf(N,1);

    % expected = max count observed across dataset
    nBC_expected = max(arrayfun(@(s) numel(getfield_safe(s,'BC',[])), TestBrake));
    nWV_expected = max(arrayfun(@(s) numel(getfield_safe(s,'WV',[])), TestBrake));

    blank = struct( ...
        'MBP_OK',false, ...
        'BC_valid',0,'BC_total',0,'BC_expected',nBC_expected,'BC_OK',false, ...
        'WV_valid',0,'WV_total',0,'WV_expected',nWV_expected,'WV_OK',false, ...
        'HasErrors',false);
    report = repmat(blank, N, 1);

    for i = 1:N
        S = TestBrake(i);

        % MBP check
        MBP_OK = has_nonempty_series(S,'MBP_Time','MBP_Pressure',10);

        % BC: SensorError==0 AND BrakingAct==1
        [BC_valid, BC_total] = count_valid_streams( ...
            getfield_safe(S,'BC',[]), ...
            'Time','Pressure','SensorError',0, 'BrakingAct',1);

        % WV: WV_SensorError==0
        [WV_valid, WV_total] = count_valid_streams( ...
            getfield_safe(S,'WV',[]), ...
            'Time','Pressure','WV_SensorError',0, '', 0);

        BC_OK = (BC_valid == nBC_expected && BC_total >= nBC_expected);
        WV_OK = (WV_valid == nWV_expected && WV_total >= nWV_expected);

        hasErr = any_true_field(S, {'SV_Error','UP_Error','EmergencyBrake'});

        if MBP_OK && BC_OK && WV_OK
            scores(i) = 100 - 10*hasErr;   % small penalty if those flags are set
        end

        report(i).MBP_OK   = MBP_OK;
        report(i).BC_valid = BC_valid; report(i).BC_total = BC_total; report(i).BC_OK = BC_OK;
        report(i).WV_valid = WV_valid; report(i).WV_total = WV_total; report(i).WV_OK = WV_OK;
        report(i).HasErrors = hasErr;
    end

    if any(isfinite(scores))
        [~, refPhase] = max(scores);
    else
        refPhase = NaN;
    end
end

% ===== helpers =====
function tf = has_nonempty_series(S, tField, pField, minLen)
    tf = false;
    if ~isfield(S,tField) || ~isfield(S,pField), return; end
    t = S.(tField); p = S.(pField);
    if isempty(t) || isempty(p), return; end
    tf = isvector(t) && isvector(p) && numel(t)==numel(p) && numel(t)>=minLen;
end

function val = getfield_safe(S, fld, defaultVal)
    if isfield(S,fld), val=S.(fld); else, val=defaultVal; end
end

function [nValid,nTotal] = count_valid_streams(A,tName,pName,reqField,reqVal,optField,optVal)
% Count streams with:
%   - nonempty Time & Pressure
%   - reqField == reqVal
%   - if optField~='', then optField == optVal when present
    if nargin<6 || isempty(optField), optField = ''; end
    if nargin<7, optVal = 0; end
    nValid=0; nTotal=0;
    if isempty(A) || ~isstruct(A), return; end
    nTotal = numel(A);
    for k=1:nTotal
        X = A(k);
        hasT = isfield(X,tName) && ~isempty(X.(tName));
        hasP = isfield(X,pName) && ~isempty(X.(pName));
        reqOK = isfield(X,reqField) && ~isempty(X.(reqField)) && all(X.(reqField)==reqVal);
        optOK = true;
        if ~isempty(optField)
            if isfield(X,optField) && ~isempty(X.(optField))
                optOK = all(X.(optField)==optVal);
            end
        end
        if hasT && hasP && reqOK && optOK
            nValid = nValid + 1;
        end
    end
end

function tf = any_true_field(S, fields)
    tf=false;
    for ii=1:numel(fields)
        f=fields{ii};
        if isfield(S,f) && ~isempty(S.(f)) && any(S.(f)~=0)
            tf=true; return;
        end
    end
end
